File: saltonFramework/src/main/java/com/salton123/app/BaseApplication.java
package com.salton123.app;

import android.Manifest;
import android.app.Application;
import android.content.Context;
import android.content.pm.PackageManager;
import android.os.Environment;
import android.os.Process;

import com.salton123.C;
import com.salton123.app.crash.SaltonCrashHandler;
import com.salton123.app.future.FutureTaskApplication;
import com.salton123.log.XLog;
import com.salton123.log.XLogConfig;
import com.salton123.model.manager.lifecycle.IActivityLifeCycle;

import java.io.File;

/**
 * User: newSalton@outlook.com
 * Date: 2019/5/9 17:50
 * ModifyTime: 17:50
 * Description:
 */
public class BaseApplication extends FutureTaskApplication {
    public static Application sInstance;

    public static <T extends Application> T getInstance() {
        return (T) sInstance;
    }

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        sInstance = this;
    }

    @Override
    public void lowPriority() {
        super.lowPriority();
        openXLog();
        openCrashHanlder();
        openLifeCycleHandler();
    }

    public void openLifeCycleHandler() {
        IActivityLifeCycle.Factory.get().init(sInstance);
    }

    public void openCrashHanlder() {
        SaltonCrashHandler.INSTANCE.init(true);
    }

    public void openXLog() {
        boolean hasPermission = checkPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE, Process.myPid(),
                Process.myUid()) == PackageManager.PERMISSION_GRANTED;
        XLog.config(new XLogConfig()
                .setWhetherToSaveLog(hasPermission)
                .setSavePath(C.BASE_PATH));
    }
}

File: saltonFramework/src/main/java/com/salton123/app/BaseApplication.kt
package com.salton123.app

import android.Manifest
import android.content.pm.PackageManager
import android.os.Environment
import android.os.Process
import android.util.Log
import com.salton123.log.XLogConfig
import com.salton123.model.manager.lifecycle.IActivityLifeCycle
import java.io.File


/**
 * User: 巫金生(newSalton@outlook.com)
 * Date: 2015-08-06
 * Time: 12:23
 * Desc:捕获应用异常Application以及完成整个应用退出；在这里进行全局变量的传递；
 * 在这里完成低内存的释放；在这里捕获未抓住的异常；用于应用配置, 预加载处理
 */
open class BaseApplication : FutureTaskApplication() {
    val TAG = "BaseApplication"
    override fun onCreate() {
        super.onCreate()
        sInstance = this
    }

    override fun onTerminate() {
        super.onTerminate()
        IActivityLifeCycle.Factory.get().unInit()
        System.gc()
    }

    override fun highPriority(): Boolean {
        return true
    }

    override fun mediumPriority(): Boolean {
        IActivityLifeCycle.Factory.get().init(sInstance)
        return true
    }

    override fun lowPriority(): Boolean {
        Log.e(TAG, "step two low:" + Thread.currentThread().name)
        openCrashHanlder()
        initXlog()
        return true
    }

    /**
     * 异常处理收集
     */
    open fun openCrashHanlder() {
        SaltonCrashHandler.isShowCrashPanel = false
    }

    /**
     * Initialize XLog.
     */
    private fun initXlog() {
        var path = File(Environment.getExternalStorageDirectory(), "salton").path
        var hasPermission = checkPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE, Process.myPid(),
            Process.myUid()) == PackageManager.PERMISSION_GRANTED
        path = path + File.separator + sInstance.packageName
        XLogConfig.init(XLogConfig.Builder()
            .whetherToSaveLog(hasPermission)
            .setSavePath(path)
            .build())
    }

    @Suppress("UNCHECKED_CAST")
    companion object {
        private lateinit var sInstance: BaseApplication
        @JvmStatic
        fun <T : BaseApplication> getInstance(): T {
            return sInstance as T
        }
    }
}
